01-algorithms/02-algorithms-divide-conquer.md

```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
+    mid = (left + right) // 2
 
-    # if the array has 0 or 1 elements, it's already sorted
-    if high - low <= 1:
-        return
+    # recur for the left sublist
+    minNum, maxNum = findMinAndMax(nums, left, mid, minNum, maxNum)
 
-    # find the midway point where we will divide
-    mid = (low + high) // 2
+    # recur for the right sublist
+    minNum, maxNum = findMinAndMax(nums, mid + 1, right, minNum, maxNum)
 
-    # apply merge sort to each half of the array
-    merge_sort(array, low, mid)
-    merge_sort(array, mid, high)
+    return minNum, maxNum
+```
+</details>
 
-    # merge the sorted left and right subarrays
-    merge(array, low, mid, high)
+### Binary Search: A modified approach
 
-def merge(array, low, mid, high):
-    merged = []
-    l = low
-    r = mid
+Binary search can be used to solve many problems related to searching through sorted data. The following problem offers an opportunity to use the binary search algorithm using a modified approach.
 
-    # continue merging by comparison until one subarray is empty
-    while l < mid and r < high:
-        # take either the left or right value
-        if array[l] <= array[r]:
-            merged.append(array[l])
-            l += 1
-        else:
-            merged.append(array[r])
-            r += 1
+```
+Suppose you are given a sorted array of non-negative distinct integers. 
+You are tasked with finding the smallest missing non-negative element inside of the array.
 
-    # if there was data remaining in the left array take the rest
-    while l < mid:
-        merged.append(array[l])
-        l += 1
+If present, the array should start with the integer 0.
 
-    # or if there was data remaining in the right array take the rest
-    while r < high:
-        merged.append(array[r])
-        r += 1
+For example:
+If the array is [0, 1, 2, 6, 9, 11, 15], the smallest missing element is 3.
 
-    # copy from the auxiliary array back to the main array
-    m = 0
-    l = low
-    while l < high:
-        array[l] = merged[m]
-        l += 1
-        m += 1
+If the array is [0, 1, 2, 3, 4, 5], the smallest missing element is 6.
 ```
 
-</details>
+This problem can be solved in O(n) time using a linear search to find the first index containing an element that does not match its value. However, this approach does not take advantage of the fact the input is sorted.
 
-## Example: Binary Search (Decrease-and-Conquer)
+We can instead use a modified binary search algorithm to solve the problem in O(log(n)) time.
 
-Binary search can be considered a decrease-and-conquer algorithm because it decreases the size of the array by half with each step. No matter how large the original array was, at each step we'll be able to discard half the remaining data from consideration! The half-sized array will then be analyzed with the same step as the full-sized array.
+Spend *no more than 15 minutes* writing a modified binary search solution to the problem detailed above.
```

Comment:
Again, I wonder if it would serve students to give them a workspace (via a challenge question) to do this work. Can be zero points so that it doesn't affect completion.

Reply:
Good idea, I'll add a challenge question for the two exercises